File: src/spade/reporter/audit/kernel-modules/spade/audit/helper/syscall/network.c
/*
 --------------------------------------------------------------------------------
 SPADE - Support for Provenance Auditing in Distributed Environments.
 Copyright (C) 2025 SRI International

 This program is free software: you can redistribute it and/or
 modify it under the terms of the GNU General Public License as
 published by the Free Software Foundation, either version 3 of the
 License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program. If not, see <http://www.gnu.org/licenses/>.
 --------------------------------------------------------------------------------
 */

#include <linux/errno.h>

#include "spade/audit/helper/syscall/network.h"

#include "spade/audit/helper/task.h"
#include "spade/audit/helper/sock.h"
#include "spade/audit/msg/network/create.h"
#include "spade/audit/global/global.h"
#include "spade/util/log/log.h"


int helper_syscall_network_copy_saddr_and_size_from_userspace(
    struct sockaddr *dst,
    uint32_t *dst_size,
    struct sockaddr __user *src,
    uint32_t __user *src_size
)
{
    const char *log_id = "helper_syscall_network_copy_saddr_and_size_from_userspace";

    int err;

    if (!dst || !dst_size || !src || !src_size)
        return -EINVAL;

    err = helper_sock_copy_sock_len_from_userspace(dst_size, src_size);
    if (err != 0)
    {
        util_log_warn(log_id, "Failed to copy sockaddr len from userspace");
        return err;
    }

    // dst_size is already properly copied above. Therefore, passing it as src_size.
    return helper_syscall_network_copy_only_saddr_from_userspace(
        dst, dst_size,
        src, *dst_size
    );
}

int helper_syscall_network_copy_only_saddr_from_userspace(
    struct sockaddr *dst,
    uint32_t *dst_size,
    struct sockaddr __user *src,
    uint32_t src_size
)
{
    const char *log_id = "helper_syscall_network_copy_only_saddr_from_userspace";

    int err;

    if (!dst || !dst_size || !src)
        return -EINVAL;

    err = helper_sock_copy_saddr_from_userspace(
        dst, sizeof(*dst),
        src, src_size
    );
    if (err != 0)
    {
        util_log_warn(log_id, "Failed to copy sockaddr from userspace");
        return err;
    }

    *dst_size = src_size;

    return 0;
}

int helper_syscall_network_get_peer_saddr_from_fd(
    struct sockaddr *dst,
    uint32_t *dst_size,
    int sockfd
)
{
    const char *log_id = "helper_syscall_network_get_peer_saddr_from_fd";
    int err;
    struct helper_sock_saddr_info saddr_info = {0};
    int peer_mode = 1; // remote
    bool include_ns_info = false;

    if (!dst || !dst_size)
        return -EINVAL;

    err = helper_sock_get_saddr_info_from_fd(
        &saddr_info, include_ns_info, peer_mode, sockfd
    );
    if (err != 0)
    {
        util_log_warn(log_id, "Failed to get remote fd saddr info");
        return err;
    }
    memcpy(dst, &(saddr_info.saddr), saddr_info.saddr_size);
    *dst_size = saddr_info.saddr_size;
    return 0;
}

int helper_syscall_network_copy_saddr_and_size_in_msghdr_from_userspace(
    struct sockaddr *dst, uint32_t *dst_size,
    struct msghdr __user *src
)
{
    const char *log_id = "helper_syscall_network_copy_saddr_and_size_in_msghdr_from_userspace";
    int err;

    if (!dst || !dst_size || !src)
        return -EINVAL;

    err = helper_sock_copy_saddr_in_msghdr_from_userspace(
        dst, dst_size, src
    );
    if (err != 0)
    {
        util_log_warn(log_id, "Failed to get copy saddr in msghdr from userspace");
        return err;
    }
    return 0;
}

int helper_syscall_network_sockfd_is_connected(
    bool *dst, int sockfd
)
{
    const char *log_id = "helper_syscall_network_sockfd_is_connected";

    int err;

    if (!dst)
        return -EINVAL;

    err = helper_sock_is_sockfd_in_connected_state(dst, sockfd);
    if (err != 0)
    {
        util_log_warn(log_id, "Failed to check if sockfd state");
        return err;
    }

    return 0;
}

int helper_syscall_network_populate_msg(
    struct msg_network *msg,
    struct kernel_syscall_context_post *sys_ctx,
    int subject_sockfd,
    struct sockaddr *remote_saddr,
    uint32_t remote_saddr_size
)
{
    const char *log_id = "helper_syscall_network_populate_msg";

    int err;

    struct helper_sock_saddr_info local_saddr_info = {0};
    int local_peer_mode = 0;

    if (!msg || !sys_ctx || !remote_saddr)
        return -EINVAL;

    err = helper_sock_get_saddr_info_from_fd(
        &local_saddr_info, global_is_network_logging_ns_info(), local_peer_mode, subject_sockfd
    );
    if (err != 0)
    {
        util_log_warn(log_id, "Failed to get local fd saddr info");
        return err;
    }

    err = helper_task_populate_process_info_from_current_task(&msg->proc_info);
    if (err != 0)
    {
        util_log_warn(log_id, "Failed to copy current process info");
        return err;
    }

    msg->fd = subject_sockfd;
    msg->local_saddr = local_saddr_info.saddr;
    msg->local_saddr_size = local_saddr_info.saddr_size;
    msg->net_ns_inum = local_saddr_info.net_ns_inum;
    msg->remote_saddr = *remote_saddr;
    msg->remote_saddr_size = remote_saddr_size;
    msg->sock_type = local_saddr_info.sock_type;
    msg->syscall_number = sys_ctx->header.sys_num;
    msg->syscall_result = sys_ctx->sys_res.ret;
    msg->syscall_success = sys_ctx->sys_res.success;

    return 0;
}

File: pkg/java/src/main/java/spade/reporter/Network.java
/*
 --------------------------------------------------------------------------------
 SPADE - Support for Provenance Auditing in Distributed Environments.
 Copyright (C) 2015 SRI International

 This program is free software: you can redistribute it and/or
 modify it under the terms of the GNU General Public License as
 published by the Free Software Foundation, either version 3 of the
 License, or (at your option) any later version.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program. If not, see <http://www.gnu.org/licenses/>.
 --------------------------------------------------------------------------------
 */
package spade.reporter;

import spade.core.AbstractReporter;
import spade.edge.opm.Used;
import spade.edge.opm.WasGeneratedBy;

import java.io.BufferedReader;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.io.PrintStream;
import java.net.InetAddress;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.LinkedHashMap;

public class Network extends AbstractReporter implements Runnable {

    PrintStream outputStream = System.out;
    PrintStream errorStream = System.err;
    java.lang.Process process;
    boolean shutdown = false;
    HashMap<String, HashSet<String>> currentPidConnectionMap = new HashMap<>();
    HashMap<String, HashSet<String>> nextPidConnectionMap = new HashMap<>();

    @Override
    public boolean launch(String arguments) {

        boolean started = false;

        try {
            new Thread(this).start();
            started = true;
        } catch (Exception exception) {
            exception.printStackTrace(errorStream);
        }
        return started;
    }

    @Override
    public void run() {

        BufferedReader lsofBufferedReader = initialize();

        if (lsofBufferedReader != null) {
            while (!shutdown) {
                scanList(lsofBufferedReader);
            }
            process.destroy();
        }
    }

    BufferedReader initialize() {

        BufferedReader lsofBufferedReader = null;

        try {
            process = Runtime.getRuntime().exec("lsof -F pn -i -r 1");
            InputStream inputStream = process.getInputStream();
            InputStreamReader inputStreamReader = new InputStreamReader(inputStream);
            lsofBufferedReader = new BufferedReader(inputStreamReader);
        } catch (Exception exception) {
            exception.printStackTrace(errorStream);
        }

        return lsofBufferedReader;
    }

    void scanList(BufferedReader lsofBufferedReader) {

        String pid;
        boolean done = false;

        do {
            pid = scanProcess(lsofBufferedReader);

            if (pid != null) {
                scanConnections(lsofBufferedReader, pid);
            }

            if (peekString(lsofBufferedReader, "m")) {

                currentPidConnectionMap = nextPidConnectionMap;
                nextPidConnectionMap = new HashMap<>();
                try {
                    lsofBufferedReader.readLine();
                } catch (Exception exception) {
                    exception.printStackTrace(errorStream);
                }
                done = true;
            }

        } while (!done);
    }

    String scanProcess(BufferedReader lsofBufferedReader) {

        String pid = null;

        try {
            HashSet<String> connectionSet;
            String line;

            line = lsofBufferedReader.readLine();

            if (line.startsWith("p")) {
                pid = line.substring(1);

                if (currentPidConnectionMap.containsKey(pid)) {
                    connectionSet = currentPidConnectionMap.get(pid);
                } else {
                    connectionSet = new HashSet<>();
                }
                nextPidConnectionMap.put(pid, connectionSet);
            } else {
                errorStream.println("Expected line to start with 'p':\n" + "\t" + line);
            }
        } catch (Exception exception) {
            exception.printStackTrace(errorStream);
        }

        return pid;
    }

    void scanConnections(BufferedReader lsofBufferedReader, String pid) {

        try {
            String line, connection;
            HashSet<String> connectionSet;
            boolean done = false;

            connectionSet = nextPidConnectionMap.get(pid);
            do {
                if (peekString(lsofBufferedReader, "n")) {

                    line = lsofBufferedReader.readLine();
                    connection = line.substring(1);

                    if (connection.matches(".*:.*->.*:.*")) {
                        if (!connectionSet.contains(connection)) {
                            connectionSet.add(connection);
                            emitOPM(pid, connection);
                        }
                    }
                } else {
                    done = true;
                }
            } while (!done);
        } catch (Exception exception) {
            exception.printStackTrace(errorStream);
        }
    }

    void emitString(String pid, String connection) {
        outputStream.println("PID: " + pid + "\t\tConnection: " + connection);
    }

    void emitOPM(String pid, String connection) {

        try {
            String source, destination, port;
            String[] endPoint, endPoints;
            LinkedHashMap<String, String> annotations;
            boolean endPointMatched = false;
            Date currentTime;

            spade.vertex.opm.Process processVertex;
            spade.vertex.opm.Artifact networkVertex;
            WasGeneratedBy wasGeneratedByEdge;
            Used usedEdge;

            // Create process vertex.
            annotations = new LinkedHashMap<>();
            annotations.put("pid", pid);
            processVertex = new spade.vertex.opm.Process();
            processVertex.addAnnotations(annotations);

            if (!putVertex(processVertex)) {
                errorStream.println("Buffer did not accept process artifact:" + "\n\t pid" + pid);
            }

            // Create network artifact.
            annotations = new LinkedHashMap<>();
            endPoints = connection.split("->");

            endPoint = endPoints[0].split(":");
            source = endPoint[0];
            annotations.put("source host", source);
            port = endPoint[1];
            annotations.put("source port", port);

            endPoint = endPoints[1].split(":");
            destination = endPoint[0];
            annotations.put("destination host", destination);
            port = endPoint[1];
            annotations.put("destination port", port);

            networkVertex = new spade.vertex.opm.Artifact();
            networkVertex.addAnnotations(annotations);

            if (!putVertex(networkVertex)) {
                errorStream.println("Buffer did not accept connection artifact:" + "\n\t " + connection);
            }

            // Create an outgoing edge.
            if (InetAddress.getByName(destination).isSiteLocalAddress()) {
                annotations = new LinkedHashMap<>();
                currentTime = new Date();
                annotations.put("time", currentTime.toString());
                usedEdge = new Used(processVertex, networkVertex);
                usedEdge.addAnnotations(annotations);
                if (!putEdge(usedEdge)) {
                    errorStream.println("Buffer did not accept outgoing "
                            + "connection edge:\n\t pid: " + pid
                            + "\n\t connection: " + connection
                            + "\n\t time: " + currentTime.toString());
                }
                endPointMatched = true;
            }

            // Create an incoming edge.
            if (InetAddress.getByName(source).isSiteLocalAddress()) {
                annotations = new LinkedHashMap<>();
                currentTime = new Date();
                annotations.put("time", currentTime.toString());
                wasGeneratedByEdge
                        = new WasGeneratedBy(networkVertex, processVertex);
                wasGeneratedByEdge.addAnnotations(annotations);
                if (!putEdge(wasGeneratedByEdge)) {
                    errorStream.println("Buffer did not accept incoming "
                            + "connection edge:\n\t pid: " + pid
                            + "\n\t connection: " + connection
                            + "\n\t time: " + currentTime.toString());
                }

                endPointMatched = true;
            }

            if (!endPointMatched) {
                errorStream.println("Neither endpoint is local:\n\t"
                        + connection);
            }
        } catch (Exception exception) {
            exception.printStackTrace(errorStream);
        }
    }

    boolean peekString(BufferedReader bufferedReader, String nextString) {

        boolean match = false;

        try {
            int charactersRead, nextStringLength = nextString.length();
            char[] peekBuffer = new char[nextStringLength];
            String peekString;

            bufferedReader.mark(nextStringLength);
            charactersRead = bufferedReader.read(peekBuffer, 0,
                    nextStringLength);
            bufferedReader.reset();

            if (charactersRead == nextStringLength) {
                peekString = new String(peekBuffer);
                if (peekString.equals(nextString)) {
                    match = true;
                }
            }
        } catch (Exception exception) {
            exception.printStackTrace(errorStream);
        }

        return match;
    }

    @Override
    public boolean shutdown() {

        shutdown = true;
        return true;
    }
}
